### PR TITLE
Purge remote translog basis the latest metadata for remote-backed indexes

### PR DESCRIPTION
**Pull request by ashking94**
### Description
During failover or relocation of primary shard, during initialisation of RemoteFsTranslog, we read remote translog metadata and download translog and checkpoint file for the corresponding primary term and generation as present in remote translog metadata file. After a flush, we clear *.tlog and *.ckp files, but the remote metadata file continues to hold references to the deleted generations. We need to handle the same.

We should delete remote translog considering the latest metadata file uploaded. The tlog/ckp files referenced by that metadata file should not be deleted . By doing this we will be able to restore translog from that metadata.

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/5845

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ashking94** [comment]
There will be a follow up PR that makes the purging of translog files on remote truly async with a dedicated threadpool. Issue - #6104.

**Review by gbbafna — COMMENTED** [review]
LGTM .

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java [hunk 1]
```diff
@@ -195,14 +195,50 @@ private TransferFileSnapshot prepareMetadata(TransferSnapshot transferSnapshot)
...
+        });
+        transferService.deleteBlobsAsync(remoteBaseTransferPath.add(String.valueOf(primaryTerm)), files, new ActionListener<>() {
+            @Override
+            public void onResponse(Void unused) {
+                fileTransferTracker.delete(files);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:223]: If checkpoint file deletion fails, it won't be tried next time, is that correct?

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 2]
```diff
@@ -327,13 +338,57 @@ protected void setMinSeqNoToKeep(long seqNo) {
...
+        deleteOlderPrimaryTranslogFilesFromRemoteStore();
+    }
+
+    private void deleteOlderPrimaryTranslogFilesFromRemoteStore() {
+        // The deletion of older translog files in remote store is on best-effort basis, there is a possibility that there
+        // are older files that are no longer needed and should be cleaned up. In here, we delete all files that are part
+        // of older primary term.
+        if (olderPrimaryCleaned.trySet(Boolean.TRUE)) {
+            logger.info("Cleaning up translog uploaded by previous primaries");
+            for (long oldPrimaryTerm = current.getPrimaryTerm() - 1; oldPrimaryTerm >= 0; oldPrimaryTerm--) {
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:372]: Shouldn't this initialized according to primary term mapped to min translog generation? If min translog generation is mapped to primary term X, oldPrimaryTerm should initialize to X - 1.

Also, do we need to go till 0 each time? This means time to execute this method will increase as primaryTerm increases. Can we identify if path does not exist as part of `onFailure` and break the loop?

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java (hunk 1 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java:223]: The deletion is on best effort basis. If the deletion fails, the deletion happens on initialisation of RemoteFsTranslog (failover/relocation).

File: server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java [hunk 3]
```diff
@@ -195,14 +195,50 @@ private TransferFileSnapshot prepareMetadata(TransferSnapshot transferSnapshot)
...
+                        generations
+                    ),
+                    e
+                );
+            }
+        });
+    }
+
+    /**
+     * Handles deletion on translog files for a particular primary term.
```
[the remark this answers is not in the document]
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java]: Incorporated.

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 4]
```diff
@@ -327,13 +338,57 @@ protected void setMinSeqNoToKeep(long seqNo) {
...
-            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), reader.generation);
-        } catch (IOException ignored) {
-            logger.error("Exception {} while deleting generation {}", ignored, reader.generation);
+            translogTransferManager.deleteTranslogAsync(primaryTermSupplier.getAsLong(), generations);
+        } catch (IOException e) {
+            logger.error(() -> new ParameterizedMessage("Exception occurred while deleting generation {}", generations), e);
+        }
+    }
+
+    public void trimUnreferencedReaders() throws IOException {
```
[the remark this answers is not in the document]
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:353]: Incorporated.

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java (hunk 2 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java]: >Shouldn't this initialized according to primary term mapped to min translog generation? If min translog generation is mapped to primary term X, oldPrimaryTerm should initialize to X - 1.

Yes. Incorporated the comment.

>Also, do we need to go till 0 each time? This means time to execute this method will increase as primaryTerm increases. Can we identify if path does not exist as part of onFailure and break the loop?

No, made changes here too. Now, I am reading all primary terms that exists in the remote store and deleting those which are smaller than the minimum referenced primary term in the latest metadata file.

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 5]
```diff
@@ -327,13 +341,87 @@ protected void setMinSeqNoToKeep(long seqNo) {
...
+    }
+
+    private Set<Long> getPrimaryTermsInRemote() {
         try {
-            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), reader.generation);
-        } catch (IOException ignored) {
-            logger.error("Exception {} while deleting generation {}", ignored, reader.generation);
+            return translogTransferManager.listPrimaryTerms();
+        } catch (IOException e) {
+            logger.error("Exception occurred while getting oldest primary term in remote store", e);
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:407]: nit: getting **all** the primary terms from the remote store

**Comment by sachinpkale** [comment]
Added some minor comments. Overall code looks good.

**Review by sachinpkale — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java (hunk 5 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java]: Incorporated

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 6]
```diff
@@ -327,13 +341,87 @@ protected void setMinSeqNoToKeep(long seqNo) {
...
+        try {
+            TranslogTransferMetadata translogMetadata = translogTransferManager.readMetadata();
+            if (translogMetadata != null && translogMetadata.getGenerationToPrimaryTermMapper().isEmpty() == false) {
+                OptionalLong min = translogMetadata.getGenerationToPrimaryTermMapper().values().stream().mapToLong(Long::valueOf).min();
+                if (min.isPresent()) {
+                    minPrimaryTerm = min.getAsLong();
+                }
+            }
+        } catch (IOException e) {
+            logger.error("Exception occurred while getting max primary term in remote translog metadata", e);
```
[the remark this answers is not in the document]
Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java]: This code is not there anymore.

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 7]
```diff
@@ -327,13 +340,73 @@ protected void setMinSeqNoToKeep(long seqNo) {
...
         try {
-            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), reader.generation);
-        } catch (IOException ignored) {
-            logger.error("Exception {} while deleting generation {}", ignored, reader.generation);
+            return translogTransferManager.listPrimaryTerms();
+        } catch (IOException e) {
+            logger.error("Exception occurred while getting primary terms from remote store", e);
         }
-        super.deleteReaderFiles(reader);
+        return LongStream.range(0, current.getPrimaryTerm()).boxed().collect(Collectors.toSet());
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:410]: This doesn't look the right behavior. We should propagate the exception and let the caller handle the same. 

I think it is okay to not delete older primary folders here.

Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:410]: @gbbafna So, this method in RemoteFsTranslog in the caller which is asking to list all primary terms in remote. This is a private method and if someone wants to get all primary terms with in RemoteFsTranslog class, then they use `translogTransferManager.listPrimaryTerms()` which throws the exception to the caller. Currently for the deletion usecase, if there are failures in getting list of primary terms from remote then we default to clearing all primary terms which are below the minimum primary term referenced in metadata file.

Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:410]: The name `getPrimaryTermsInRemote` doesn't suggest that it is going to give all previous primary terms in remote translog store. Should we just inline it then in `deleteOlderPrimaryTranslogFilesFromRemoteStore` as whatever it is returning is strictly meant for  `deleteOlderPrimaryTranslogFilesFromRemoteStore`?

Comment (ashking94) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java]: Sounds reasonable. Incorporated it.

**Review by gbbafna — APPROVED** [review]

**Comment by gbbafna** [comment]
Build failed due to 

```
> Task :modules:ingest-common:jarHell FAILED
| Output for C:\hostedtoolcache\windows\Java_Adopt_jdk\11.0.18-10\x64\bin\java.exe:#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (malloc) failed to allocate 57616 bytes for Chunk::new
# An error report file with more information is saved as:
# D:\a\OpenSearch\OpenSearch\modules\ingest-common\hs_err_pid6884.log
````

**Comment by ashking94** [comment]
Build succeeded. @gbbafna

**Pull request merged**